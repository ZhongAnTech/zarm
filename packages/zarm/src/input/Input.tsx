import * as React from 'react';
import classnames from 'classnames';
import { CloseCircleFill } from '@zarm-design/icons';
import { getValue } from './utils';
import { ConfigContext } from '../n-config-provider';
import type { BaseInputTextProps, BaseInputTextareaProps } from './interface';

const regexAstralSymbols = /[\uD800-\uDBFF][\uDC00-\uDFFF]|\n/g;

const countSymbols = (text = '') => {
  return text.replace(regexAstralSymbols, '_').length;
};

export interface InputTextProps
  extends BaseInputTextProps,
    React.InputHTMLAttributes<HTMLInputElement> {
  prefixCls?: string;
}

export type InputTextareaProps = BaseInputTextareaProps &
  React.TextareaHTMLAttributes<HTMLTextAreaElement>;

export type InputProps = {
  type?: string;
} & (InputTextProps | InputTextareaProps);

const Input = React.forwardRef<unknown, InputProps>((props, ref) => {
  const {
    type,
    disabled,
    autoFocus,
    readOnly,
    clearable,
    showLength,
    autoHeight,
    rows,
    className,
    style,
    value,
    defaultValue,
    maxLength,
    onChange,
    onBlur,
    onFocus,
    ...restProps
  } = props as InputTextProps & InputTextareaProps;

  const wrapperRef = (ref as any) || React.createRef<HTMLDivElement>();
  const inputRef = React.useRef<HTMLTextAreaElement | HTMLInputElement>();

  const [currentValue, setCurrentValue] = React.useState(getValue({ value, defaultValue }));
  const [focused, setFocused] = React.useState<boolean>(autoFocus!);

  const isTextarea = type === 'text' && 'rows' in props;
  let blurFromClear = false;
  let blurTimeout: number;

  const showClearIcon =
    clearable &&
    !readOnly &&
    !disabled &&
    typeof value !== 'undefined' &&
    typeof onChange !== 'undefined' &&
    currentValue.length > 0;

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-input`;

  const cls = classnames(prefixCls, className, {
    [`${prefixCls}--textarea`]: isTextarea,
    [`${prefixCls}--disabled`]: disabled,
    [`${prefixCls}--clearable`]: showClearIcon,
    [`${prefixCls}--readonly`]: readOnly,
    [`${prefixCls}--focus`]: focused,
  });

  const focus = () => {
    inputRef.current!.focus();
  };

  const blur = () => {
    inputRef.current!.blur();
  };

  const onInputBlur = (e) => {
    blurTimeout = window.setTimeout(() => {
      if (!blurFromClear && document.activeElement !== inputRef.current) {
        setFocused(false);

        if (typeof onBlur === 'function') {
          onBlur(e);
        }
      }
      blurFromClear = false;
    }, 200);
  };

  const onInputFocus = (e) => {
    setFocused(true);

    if (typeof onFocus === 'function') {
      onFocus(e);
    }
  };

  const onInputChange = (e) => {
    setCurrentValue(e.target.value);

    if (typeof onChange === 'function') {
      onChange(e);
    }
  };

  const onInputClear = (e) => {
    blurFromClear = true;
    setCurrentValue('');
    focus();

    if (typeof onChange !== 'function') {
      return;
    }

    const event = Object.create(e);
    const target = inputRef.current!;
    const originalValue = target.value;

    event.target = target;
    event.currentTarget = target;
    target.value = '';
    onChange(event);
    // reset target ref value
    target.value = originalValue;
  };

  const commonProps = {
    ...restProps,
    maxLength,
    disabled,
    autoFocus,
    readOnly,
    value: 'value' in props ? currentValue : undefined,
    defaultValue: 'defaultValue' in props ? defaultValue : undefined,
    onChange: 'onChange' in props ? onInputChange : undefined,
    onFocus: onInputFocus,
    onBlur: onInputBlur,
  };

  // 渲染输入框
  const renderInput = isTextarea ? (
    <textarea
      ref={inputRef as React.RefObject<HTMLTextAreaElement>}
      {...(commonProps as React.HTMLAttributes<HTMLTextAreaElement>)}
      rows={rows}
    />
  ) : (
    <input
      ref={inputRef as React.RefObject<HTMLInputElement>}
      {...(commonProps as React.HTMLAttributes<HTMLInputElement>)}
      type={type}
    />
  );

  // 渲染文本内容
  const renderText = <div className={`${prefixCls}__content`}>{currentValue}</div>;

  // 渲染文字长度
  const renderTextLength = showLength && maxLength && (
    <div className={`${prefixCls}__length`}>{`${countSymbols(currentValue)}/${maxLength}`}</div>
  );

  // 渲染清除按钮
  const renderClearIcon = showClearIcon && (
    <CloseCircleFill className={`${prefixCls}__clear`} onClick={onInputClear} />
  );

  React.useImperativeHandle(wrapperRef, () => ({
    focus,
    blur,
  }));

  React.useEffect(() => {
    setCurrentValue(getValue({ value, defaultValue }));
  }, [value, defaultValue]);

  React.useEffect(() => {
    if (!autoHeight) return;

    inputRef.current!.style.height = 'auto';
    inputRef.current!.scrollTop = 0;
    inputRef.current!.style.height = `${inputRef.current!.scrollHeight}px`;

    if (readOnly && rows) {
      inputRef.current!.style.height = `${inputRef.current!.scrollHeight * rows}px`;
    }
  }, [autoHeight, readOnly, rows, value]);

  React.useEffect(() => {
    return () => {
      !!blurTimeout && clearTimeout(blurTimeout);
    };
  });

  return (
    <div ref={wrapperRef} className={cls} style={style}>
      {!readOnly ? renderInput : renderText}
      {renderTextLength}
      {renderClearIcon}
    </div>
  );
});

Input.displayName = 'Input';

Input.defaultProps = {
  type: 'text',
  disabled: false,
  autoFocus: false,
  readOnly: false,
  clearable: true,
  showLength: false,
  autoHeight: false,
};

export default Input;
