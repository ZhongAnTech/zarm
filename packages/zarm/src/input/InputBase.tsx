import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { CloseCircleFill } from '@zarm-design/icons';
import { InputBaseProps } from './PropsType';
import { getValue } from './utils';

export default class InputBase extends PureComponent<InputBaseProps, any> {
  static defaultProps: InputBaseProps = {
    prefixCls: 'za-input',
    disabled: false,
    type: 'text',
    clearable: true,
    readOnly: false,
  };

  private input;

  private onBlurTimeout;

  private blurFromClear;

  constructor(props) {
    super(props);
    this.state = {
      focused: props.autoFocus || false,
      value: getValue(props),
    };
  }

  componentDidMount() {
    const { focused } = this.state;

    if (focused) {
      this.input.focus();
    }
  }

  static getDerivedStateFromProps(nextProps) {
    if ('value' in nextProps) {
      return {
        value: getValue(nextProps),
      };
    }
    return null;
  }

  componentWillUnmount() {
    if (this.onBlurTimeout) {
      clearTimeout(this.onBlurTimeout);
      this.onBlurTimeout = null;
    }
  }

  onFocus = (e) => {
    this.setState({
      focused: true,
    });

    const { onFocus } = this.props;
    if (typeof onFocus === 'function') {
      onFocus(e.target.value);
    }
  };

  onBlur = (e) => {
    const { onBlur } = this.props;
    const { value } = e.target;
    this.onBlurTimeout = setTimeout(() => {
      if (!this.blurFromClear && document.activeElement !== this.input) {
        this.setState({
          focused: false,
        });

        if (typeof onBlur === 'function') {
          onBlur(value);
        }
      }
      this.blurFromClear = false;
    }, 200);
  };

  onChange = (e) => {
    const { onChange } = this.props;
    const { focused } = this.state;
    const { value } = e.target;

    if (!focused) {
      this.setState({ focused: true });
    }

    if (!('value' in this.props)) {
      this.setState({ value });
    }

    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  onClear = () => {
    const { isOnComposition } = this.state;
    const { onChange, onClear } = this.props;

    this.blurFromClear = true;
    this.setState({
      value: '',
    });

    !isOnComposition && this.focus();
    typeof onChange === 'function' && onChange('');
    typeof onClear === 'function' && onClear('');
  };

  handleComposition = (e) => {
    const { onCompositionStart, onCompositionUpdate, onCompositionEnd, onChange } = this.props;

    if (e.type === 'compositionstart') {
      this.setState({
        isOnComposition: true,
      });
      if (typeof onCompositionStart === 'function') {
        onCompositionStart(e);
      }
    }

    if (e.type === 'compositionupdate') {
      if (typeof onCompositionUpdate === 'function') {
        onCompositionUpdate(e);
      }
    }

    if (e.type === 'compositionend') {
      const { value } = e.target;
      // composition is end
      this.setState({
        isOnComposition: false,
      });
      if (typeof onCompositionEnd === 'function') {
        onCompositionEnd(e);
      }
      if (typeof onChange === 'function') {
        onChange(value);
      }
    }
  };

  focus = () => {
    this.input.focus();
  };

  blur = () => {
    this.input.blur();
  };

  render() {
    const {
      prefixCls,
      className,
      disabled,
      clearable,
      readOnly,
      type,
      onClear,
      ...rest
    } = this.props;

    const { value, focused } = this.state;
    const showClearIcon = clearable && 'value' in this.props && 'onChange' in this.props;

    const cls = classnames(prefixCls, className, {
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--focus`]: focused && value && value.length > 0,
      [`${prefixCls}--clearable`]: showClearIcon,
      [`${prefixCls}--readonly`]: readOnly,
    });

    const renderInput = (
      <input
        {...rest}
        value={'value' in this.props ? value : undefined}
        autoComplete="off"
        ref={(ele) => {
          this.input = ele;
        }}
        type={type}
        disabled={disabled}
        onChange={this.onChange}
        onBlur={this.onBlur}
        onFocus={this.onFocus}
        onCompositionStart={(e) => {
          this.handleComposition(e);
        }}
        onCompositionUpdate={(e) => {
          this.handleComposition(e);
        }}
        onCompositionEnd={(e) => {
          this.handleComposition(e);
        }}
      />
    );

    const renderText = <div className={`${prefixCls}__content`}>{value}</div>;

    // clear icon
    const renderClearIcon = showClearIcon && (
      <CloseCircleFill
        className={`${prefixCls}__clear`}
        onClick={() => {
          this.onClear();
        }}
      />
    );

    return (
      <div className={cls}>
        {readOnly ? renderText : renderInput}
        {renderClearIcon}
      </div>
    );
  }
}
