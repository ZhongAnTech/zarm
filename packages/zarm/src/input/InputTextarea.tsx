import React, { PureComponent } from 'react';
import classnames from 'classnames';
import { InputTextareaProps } from './PropsType';
import { getValue } from './utils';

const regexAstralSymbols = /[\uD800-\uDBFF][\uDC00-\uDFFF]|\n/g;

const countSymbols = (text = '') => {
  return text.replace(regexAstralSymbols, '_').length;
};

export default class InputTextarea extends PureComponent<InputTextareaProps, any> {
  static defaultProps: InputTextareaProps = {
    prefixCls: 'za-input',
    disabled: false,
    autoHeight: false,
    showLength: false,
    focused: false,
  };

  private input;

  constructor(props) {
    super(props);
    this.state = {
      value: getValue(props),
      focused: props.focused || false,
    };
  }

  componentDidMount() {
    const { autoFocus } = this.props;
    const { focused } = this.state;
    if (autoFocus || focused) {
      this.input.focus();
    }
  }

  static getDerivedStateFromProps(nextProps) {
    if ('focused' in nextProps || 'value' in nextProps) {
      return {
        value: getValue(nextProps),
        focused: nextProps.focused || false,
      };
    }
    return null;
  }

  componentDidUpdate() {
    const { autoHeight, rows, readOnly } = this.props;
    const { focused } = this.state;

    if (autoHeight) {
      this.input.style.height = 'auto';
      this.input.scrollTop = 0;
      this.input.style.height = `${this.input.scrollHeight}px`;
    }
    if (autoHeight && readOnly && rows) {
      this.input.style.height = `${this.input.scrollHeight * rows}px`;
    }
    if (focused) {
      this.input.focus();
    }
  }

  onFocus = (e) => {
    if (!('focused' in this.props)) {
      this.setState({
        focused: true,
      });
    }

    const { onFocus } = this.props;
    if (typeof onFocus === 'function') {
      onFocus(e.target.value);
    }
  };

  onBlur = (e) => {
    if (!('focused' in this.props)) {
      this.setState({
        focused: false,
      });
    }
    const { onBlur } = this.props;
    if (typeof onBlur === 'function') {
      onBlur(e.target.value);
    }
  };

  handleComposition = (e) => {
    const { onCompositionStart, onCompositionUpdate, onCompositionEnd, onChange } = this.props;

    if (e.type === 'compositionstart') {
      // this.setState({
      //   isOnComposition: true,
      // });
      if (typeof onCompositionStart === 'function') {
        onCompositionStart(e);
      }
    }

    if (e.type === 'compositionupdate') {
      if (typeof onCompositionUpdate === 'function') {
        onCompositionUpdate(e);
      }
    }

    if (e.type === 'compositionend') {
      // composition is end
      // this.setState({
      //   isOnComposition: false,
      // });
      const { value } = e.target;
      if (typeof onCompositionEnd === 'function') {
        onCompositionEnd(e);
      }
      if (typeof onChange === 'function') {
        onChange(value);
      }
    }
  };

  onChange = (e) => {
    const { onChange } = this.props;
    const { value } = e.target;

    if (!('value' in this.props)) {
      this.setState({ value });
    }

    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  render() {
    const {
      prefixCls,
      className,
      maxLength,
      disabled,
      readOnly,
      autoHeight,
      showLength,
      focused,
      ...rest
    } = this.props;

    const cls = classnames(prefixCls, `${prefixCls}--textarea`, className, {
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--readonly`]: readOnly,
    });

    const { value } = this.state;
    const length = countSymbols(value);

    const textLengthRender = showLength && maxLength && (
      <div className={`${prefixCls}__length`}>{`${length}/${maxLength}`}</div>
    );

    const renderInput = (
      <textarea
        {...rest}
        value={'value' in this.props ? value : undefined}
        ref={(ele) => {
          this.input = ele;
        }}
        maxLength={maxLength}
        disabled={disabled}
        onChange={this.onChange}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        onCompositionStart={(e) => {
          this.handleComposition(e);
        }}
        onCompositionUpdate={(e) => {
          this.handleComposition(e);
        }}
        onCompositionEnd={(e) => {
          this.handleComposition(e);
        }}
      />
    );

    const renderText = (
      <div
        className={`${prefixCls}__content`}
        ref={(ele) => {
          this.input = ele;
        }}
      >
        {value}
      </div>
    );

    return (
      <div className={cls}>
        {!readOnly ? renderInput : renderText}
        {textLengthRender}
      </div>
    );
  }
}
