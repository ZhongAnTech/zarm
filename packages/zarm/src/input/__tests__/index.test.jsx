import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Input from '../index';
import InputBase from '../InputBase';
import InputTextarea from '../InputTextarea';

describe('Input', () => {
  it('renders correctly', () => {
    const wrapper = render(<Input type="text" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly if type=text and props includes rows', () => {
    const wrapper = render(<Input type="text" rows={1} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly if type isn't valid", () => {
    const wrapper = render(<Input type="xxx" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('showLength', () => {
    const wrapper = render(<Input showLength maxLength={100} type="text" rows={4} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('focus manual called correctly', () => {
    const onFocus = jest.fn();
    const id = String(Math.random());
    const wrapper = mount(<Input id={id} type="text" onFocus={onFocus} />);
    const instance = wrapper.instance();
    instance.focus();
    expect(wrapper.props().id).toEqual(id);
  });

  it('blur manual called correctly', () => {
    const onBlur = jest.fn();
    const wrapper = mount(<Input type="number" onBlur={onBlur} />);
    const instance = wrapper.instance();
    instance.focus();
    instance.blur();
    expect(onBlur).toBeCalled();
  });

  it('renders onFocus called correctly', () => {
    const onFocus = jest.fn();
    const wrapper = mount(<Input onFocus={onFocus} />);
    wrapper.find('input[type="text"]').simulate('focus');
    expect(onFocus).toBeCalled();
    // expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders onBlur called correctly', () => {
    const wrapper = mount(<InputBase />);
    const spy = jest.spyOn(wrapper.instance(), 'onBlur');
    wrapper.find('input').simulate('focus');
    wrapper.find('input').simulate('blur');
    expect(spy).toHaveBeenCalled();
  });

  it('renders onClear called correctly', () => {
    const onClear = jest.fn();
    const onChange = jest.fn();
    const wrapper = mount(<Input clearable value="" onClear={onClear} onChange={onChange} />);

    const input = wrapper.find('input[type="text"]');
    input.simulate('change', { target: { value: 'My new value' } });
    wrapper.find('i.za-input__clear').simulate('click');
    expect(onClear).toHaveBeenCalled();
    expect(input.instance().value).toEqual('');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('renders cn', () => {
  //   const onFocus = jest.fn();
  //   const wrapper = mount(
  //     <Input
  //       value=""
  //       onFocus={onFocus}
  //     />
  //   );

  //   const input = wrapper.find('input[type="text"]');
  //   input.simulate('change', { target: { value: '测试' } });
  //   wrapper.find('i.za-input__clear').simulate('click');
  //   expect(input.instance().value).toEqual('');
  //   expect(onFocus).toHaveBeenCalled();
  //   expect(toJson(wrapper)).toMatchSnapshot();
  // });
});

describe('Input.Base', () => {
  it('auto focus', () => {
    const id = String(Math.random());
    const wrapper = mount(<InputBase id={id} autoFocus />);
    expect(wrapper.props().id).toEqual(id);
  });

  it('value update', () => {
    const wrapper = mount(<InputBase value={1} />);
    wrapper.setProps({ value: 2 });
    expect(wrapper.state('value')).toBe(2);
  });

  it('composition', () => {
    const onCompositionStart = jest.fn();
    const onCompositionUpdate = jest.fn();
    const onCompositionEnd = jest.fn();
    const onChange = jest.fn();
    const wrapper = mount(
      <InputBase
        onCompositionStart={onCompositionStart}
        onCompositionUpdate={onCompositionUpdate}
        onCompositionEnd={onCompositionEnd}
        onChange={onChange}
      />,
    );
    const input = wrapper.find('input');
    input.simulate('compositionstart');
    expect(onCompositionStart).toBeCalled();
    expect(wrapper.state('isOnComposition')).toBe(true);
    expect(onChange).not.toBeCalled();
    input.simulate('compositionupdate');
    expect(onCompositionUpdate).toBeCalled();
    expect(onChange).not.toBeCalled();
    expect(wrapper.state('isOnComposition')).toBe(true);
    input.simulate('compositionend');
    expect(onCompositionEnd).toBeCalled();
    expect(onChange).toBeCalled();
    expect(wrapper.state('isOnComposition')).toBe(false);
  });
});

describe('Input.Textarea', () => {
  it('renders correctly', () => {
    const wrapper = render(<Input type="text" rows={4} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('autoHeight', () => {
    jest.useFakeTimers();
    const props = {
      autoHeight: true,
      type: 'text',
      rows: 4,
      value: 'foo',
      onChange: jest.fn(),
    };
    const wrapper = mount(<Input {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('textarea').simulate('change', { target: { value: 'this is a test!' } });
    expect(props.onChange).toBeCalledWith('this is a test!');
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('renders onFocus called correctly', () => {
    const onFocus = jest.fn();
    const wrapper = mount(<Input type="text" rows={2} onFocus={onFocus} />);
    wrapper.find('textarea').simulate('focus');
    expect(onFocus).toBeCalled();
    // expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders onBlur called correctly', () => {
    const onBlur = jest.fn();
    const wrapper = mount(<Input type="text" rows={2} onBlur={onBlur} />);
    // const spy = jest.spyOn(wrapper.instance(), 'onBlur');
    wrapper.find('textarea').simulate('focus');
    wrapper.find('textarea').simulate('blur');
    expect(onBlur).toHaveBeenCalled();
  });

  it('composition', () => {
    const onCompositionStart = jest.fn();
    const onCompositionUpdate = jest.fn();
    const onCompositionEnd = jest.fn();
    const onChange = jest.fn();
    const wrapper = mount(
      <InputTextarea
        onCompositionStart={onCompositionStart}
        onCompositionUpdate={onCompositionUpdate}
        onCompositionEnd={onCompositionEnd}
        onChange={onChange}
      />,
    );
    const textarea = wrapper.find('textarea');
    textarea.simulate('compositionstart');
    expect(onCompositionStart).toBeCalled();
    // expect(wrapper.state('isOnComposition')).toBe(true);
    expect(onChange).not.toBeCalled();
    textarea.simulate('compositionupdate');
    expect(onCompositionUpdate).toBeCalled();
    expect(onChange).not.toBeCalled();
    // expect(wrapper.state('isOnComposition')).toBe(true);
    textarea.simulate('compositionend');
    expect(onCompositionEnd).toBeCalled();
    expect(onChange).toBeCalled();
  });
});

describe('Input.Number', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Input type="number" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders onFocus called correctly', () => {
    const onFocus = jest.fn();
    const wrapper = mount(<Input type="number" onFocus={onFocus} />);
    wrapper.find('.za-input__content').simulate('click');
    expect(onFocus).toBeCalled();
    // expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });

  it('renders onClear called correctly', () => {
    const onClear = jest.fn();
    const onChange = jest.fn();
    const wrapper = mount(
      <Input clearable type="number" value="" onClear={onClear} onChange={onChange} />,
    );

    const input = wrapper.find('input[type="hidden"]');
    input.simulate('change', { target: { value: 'My new value' } });
    setTimeout(() => {
      wrapper.find('i.za-input__clear').simulate('click');
      expect(onClear).toHaveBeenCalled();
      expect(input.instance().value).toEqual('');
      expect(toJson(wrapper)).toMatchSnapshot();
    }, 3000);
  });

  it('inputNumber value 0', () => {
    const wrapper = mount(<Input clearable type="number" value={0} />);

    const input = wrapper.find('input[type="hidden"]');
    expect(input.instance().value).toEqual('0');
  });

  it('inputBase value 0', () => {
    const wrapper = mount(<Input clearable type="text" value={0} />);

    const input = wrapper.find('input[type="text"]');
    expect(input.instance().value).toEqual('0');
  });

  it('enter number', () => {
    const onChange = jest.fn();
    const wrapper = mount(<Input type="number" focused onChange={onChange} />);
    wrapper.find('input').simulate('focus');
    const keys = wrapper.find('.za-keyboard__keys');
    keys.childAt(0).simulate('click');
    expect(onChange).toBeCalledWith('1');
    wrapper.unmount();
  });

  it('input number hidden', () => {
    const onBlur = jest.fn();
    const wrapper = mount(<Input type="number" focused onBlur={onBlur} />);
    wrapper.find('input').simulate('focus');
    const keys = wrapper.find('.za-keyboard__keys');
    keys.childAt(11).simulate('click');
    expect(onBlur).toBeCalled();
    wrapper.unmount();
  });

  it('input number hidden', () => {
    const map = {};
    document.body.addEventListener = jest.fn((event, cb) => {
      map[event] = cb;
    });
    const onBlur = jest.fn();
    const wrapper = mount(<Input type="number" focused onBlur={onBlur} />);
    map.click({
      target: document.body,
    });
    expect(onBlur).toBeCalled();
    wrapper.unmount();
  });
});

describe('Input.Price', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Input type="price" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('Input.Idcard', () => {
  it('renders correctly', () => {
    const wrapper = mount(<Input type="idcard" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
