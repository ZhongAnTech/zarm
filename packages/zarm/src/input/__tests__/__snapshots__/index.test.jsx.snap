// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<div
  class="za-input"
>
  <input
    type="text"
  />
</div>
`;

exports[`Input renders correctly if type isn't valid 1`] = `
<div
  class="za-input"
>
  <input
    type="xxx"
  />
</div>
`;

exports[`Input renders correctly if type=text and props includes rows 1`] = `
<div
  class="za-input za-input--textarea"
>
  <div
    class="za-input__inner"
  >
    <textarea
      rows="1"
    />
  </div>
</div>
`;

exports[`Input showLength 1`] = `
<div
  class="za-input za-input--textarea"
>
  <div
    class="za-input__inner"
  >
    <textarea
      maxlength="100"
      rows="4"
    />
    <div
      class="za-input__length"
    >
      0/100
    </div>
  </div>
</div>
`;

exports[`Input.Textarea renders correctly 1`] = `
<div
  class="za-input za-input--textarea"
>
  <div
    class="za-input__inner"
  >
    <textarea
      rows="4"
    />
  </div>
</div>
`;
